src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilConverter.cs

```diff
@@ -0,0 +1,341 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+using System.Collections.Immutable;
+using System.Reflection.PortableExecutable;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.NET.WebAssembly.Webcil;
+
+/// <summary>
+/// Reads a .NET assembly in a normal PE COFF file and writes it out as a Webcil file
+/// </summary>
+public class WebcilConverter
+{
+
+    // Interesting stuff we've learned about the input PE file
+    public record PEFileInfo(
+        // The sections in the PE file
+        ImmutableArray<SectionHeader> SectionHeaders,
+        // The location of the debug directory entries
+        DirectoryEntry DebugTableDirectory,
+        // The file offset of the sections, following the section directory
+        FilePosition SectionStart,
+        // The debug directory entries
+        ImmutableArray<DebugDirectoryEntry> DebugDirectoryEntries
+        );
+
+    // Intersting stuff we know about the webcil file we're writing
+    public record WCFileInfo(
+        // The header of the webcil file
+        WebcilHeader Header,
+        // The section directory of the webcil file
+        ImmutableArray<WebcilSectionHeader> SectionHeaders,
+        // The file offset of the sections, following the section directory
+        FilePosition SectionStart
+    );
+
+    private readonly string _inputPath;
+    private readonly string _outputPath;
+
+    private WebcilConverter(string inputPath, string outputPath)
+    {
+        _inputPath = inputPath;
+        _outputPath = outputPath;
+    }
+
+    public static WebcilConverter FromPortableExecutable(string inputPath, string outputPath)
+    {
+        return new WebcilConverter(inputPath, outputPath);
+    }
+
+    public void ConvertToWebcil()
+    {
+        using var inputStream = File.Open(_inputPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
```

Comment:
`FileShare`  is setting what other Streams (incl in other processes, AFAICT) can do with this file.  ReadWrite is the most permissive.  When I had something else here, I had build failures because I think MSBuild and the WasmAppBuilder task were all trying to access the same input file for various reasons.

It does seem suspicious that other streams could be writing to the same file while we try to read it, but it might just be that the other access is done but just didn't close the stream yet. Something like that.